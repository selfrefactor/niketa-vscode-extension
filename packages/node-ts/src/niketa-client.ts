import {
  cleanAngularLog,
  defaultEmit,
  ERROR_ICON,
  extractNumber,
  isLintable,
  isMessageCorrect,
  JEST_BIN,
  LONG_SEPARATOR,
  maybeWarn,
  parse,
  SHORT_SEPARATOR,
  SUCCESS_ICON,
  toNumber,
} from './utils/common'
import execa from 'execa'
import {existsSync} from 'fs'
import {log} from 'helpers-fn'
import {lintFn} from 'lint-fn'
import {createServer} from 'net'
import {delay, glue, takeLast, remove, tryCatchAsync} from 'rambdax'
import {isLintOnlyMode, lintOnlyMode} from './modules/lint-only-mode'
import {createFileKey} from './utils/create-file-key'
import {getCoveragePath} from './utils/get-coverage-path'
import {getSpecFile} from './utils/get-spec-file'
import {getNewDecorations} from './utils/get-new-decorations'
import {getUncoveredMessage} from './utils/get-uncovered-message'
import {Message, JestSuccessMessage, ParseCoverage, NiketaClientInput} from './interfaces'

const EXTENDED_LOG = false

const FUNCTIONS = '☈' // 🕸
const STALE_SEPARATOR = '☄' // 🌰
const SEPARATOR = ' ' // 🎑 🔘 🧱
const STATEMENTS = '✍'
const BRANCHES = '🎋'
const LINES = '📜'

function logJest(execResult: any, enabled: boolean): void {
  if (!enabled) return

  process.stderr.write('\n🐬\n' + execResult.stderr + '\n🐬\n')
  process.stderr.write('\n🍵\n' + execResult.stdout + '\n🍵\n')
}

function fileInfo(x: string) {
  const [firstFolder, fileName] = takeLast(2, x.split('/'))

  return `${firstFolder}/${fileName}`
}

function onWrongIncomingMessage(message: Buffer) {
  console.log({
    message: message.toString(),
    type: typeof message,
  })

  return log('Error while parsing messageFromVSCode', 'error')
}

function debugLog(toLog: any, label = '') {
  if (!EXTENDED_LOG) return

  console.log(label, SHORT_SEPARATOR)
  console.log(LONG_SEPARATOR)
  console.log(toLog)
  console.log(LONG_SEPARATOR)
}

const baseEmitProps = {
  secondBarMessage: '',
  thirdBarMessage: '',
  hasDecorations: false,
  newDecorations: {
    correct: null,
    logData: '',
  }
}

export class NiketaClient {
  port: number
  testing: boolean
  coverageHolder: Record<string, any>
  emit: (x: any) => void
  initialized: boolean
  server: any
  jestChild: any
  pytestChild: any

  constructor(input: NiketaClientInput) {
    this.port = input.port
    this.testing = Boolean(input.testing)
    this.coverageHolder = {}
    this.emit = input.emit === undefined ? defaultEmit : input.emit
    this.initialized = false;
  }

  async onPytestMessage(message: Message){
    const { fileName, dir } = message
    
    try {
      const command = [
        `pipenv`,
        'run',
        `pytest`,
        fileName,
        '-v',
        '-rf',
        '--capture=no',
      ].join(' ')
      this.pytestChild = execa.command(command, {cwd: dir})
      log('sepx')
      log('Pytest start', 'info')
      const {stdout, stderr} = await this.pytestChild

      console.log(`stderr`, stderr)
      console.log(`stdout`, stdout)
      log('Pytest end', 'info')
      this.pytestChild = undefined
      this.emit({
        ...baseEmitProps,
        firstBarMessage: `SUCCESS - ${fileName}`,
        tooltip: stdout
      })
    } catch (e) {
      console.log(e, `pytest try.catch`)
      this.emit({
        ...baseEmitProps,
        firstBarMessage: 'FAILED',
      })
    }
  }
  async onJestMessage(message: Message){
    const { fileName, dir, hasTypescript, requestLintFile } = message

    if (!isMessageCorrect(message)){
      return this.emtpyAnswer(fileName, 'message is incorrect')
    }

    const lintOnly = isLintOnlyMode(fileName)
    const lintMessage = ` ${ fileInfo(fileName) }`

    /*
      Jest is setup for one of the two so we shouldn't check for both
    */
    const allowedSpecExtension = hasTypescript ? '.ts' : '.js'
    const { hasValidSpec, specFile } = getSpecFile(fileName,
      allowedSpecExtension)

    debugLog({
      lintOnly,
      hasValidSpec,
    })

    if (requestLintFile){
      debugLog('requestLintFile')

      return this.handleRequestLint({
        fileName,
        lintOnly,
        dir,
        lintMessage,
      })
    }

    if (lintOnly){
      debugLog('lintOnly')
      await lintOnlyMode(fileName)

      return this.lintAnswer(lintMessage)
    }

    if (!hasValidSpec) return this.emtpyAnswer(fileName, 'spec is not valid')

    const [
      failure,
      execResult,
      actualFileName,
      extension,
    ] = await this.execJest({
      dir,
      fileName     : fileName,
      specFileName : specFile,
    })

    if (failure){
      return this.emtpyAnswer(fileName,
        'Jest stopped for known or unknown reasons')
    }

    logJest(execResult, !this.testing)

    return this.onJestSuccess({
      specFile,
      dir,
      execResult,
      actualFileName,
      fileName,
      extension,
      hasTypescript,
    })
  }

  async handleRequestLint(input: {fileName: string, lintOnly: boolean, lintMessage: string, dir: string}) {
    const {fileName, lintOnly, lintMessage} = input
    
    if (!lintOnly && !isLintable(fileName)){
      return this.emtpyAnswer(fileName, '!lintable')
    }

    if (lintOnly){
      await lintOnlyMode(fileName)
      return this.lintAnswer(lintMessage)
    } 
    try {
      await this.applyLint(fileName)
      return this.lintAnswer(lintMessage)
    } catch (_) {
      return this.lintAnswer(`${ERROR_ICON} ${lintMessage}`)
    }
  }

  emtpyAnswer(fileName: string, reason: string) {
    debugLog(reason)
    this.emit({
      firstBarMessage: 'NO ACTION',
      secondBarMessage: undefined,
      thirdBarMessage: `${reason} - ${fileInfo(fileName)}`,
      hasDecorations: false,
    })
  }

  lintAnswer(lintMessage: string) {
    this.emit({
      firstBarMessage: 'LINT COMPLETED',
      secondBarMessage: undefined,
      thirdBarMessage: lintMessage,
      hasDecorations: false,
    })
  }

  async applyLint(fileName: string) {
    if (!existsSync(fileName)) return log(`${fileName} is deleted`, 'error')
    log('sep')
    log(`willLint ${fileName}`, 'info')
    
    // not ideal but it works for work-related TS files
    // await tryCatchAsync(lintFn, null)(fileName)

    // usual script
    const lintResult = await lintFn({
      filePath:fileName,
      prettierSpecialCase: 'local',
      cwdOverride: false,
      forceTypescript: true,
      debug:false
    })
    // console.log(lintResult, `lintResult`)
    // await tryCatchAsync(async x => lintFn(x, 'local', false, true), null)(fileName)

    log(`willLint ${fileName}`, 'success')
    log('sep')
  }

  onJestSuccess(input: JestSuccessMessage) {
    const {
      execResult,
      specFile,
      actualFileName,
      fileName,
      dir,
      extension,
      hasTypescript,
    } = input

    const hasError =
      execResult.stderr.startsWith('FAIL') ||
      execResult.stderr.includes('ERROR:')

    const {pass, message, uncovered} = this.parseCoverage({
      execResult,
      hasError,
      actualFileName,
      fileName,
      extension,
    })
    const {newDecorations, hasDecorations} = getNewDecorations({
      execResult,
      fileName,
      hasTypescript,
    })
    const firstBarMessage = pass ? message : ERROR_ICON
    const secondBarMessage = getUncoveredMessage(uncovered)

    debugLog(
      {
        pass,
        hasTypescript,
        message,
        hasError,
        newDecorations,
        hasDecorations,
      },
      'vscode.message'
    )
    const shorterSpecFile = remove(dir, specFile)
    
    this.emit({
      firstBarMessage,
      secondBarMessage,
      thirdBarMessage: shorterSpecFile,
      hasDecorations,
      newDecorations,
    })
  }

  async execJest(input: {fileName: string, dir: string, specFileName: string}) {
    const {fileName, dir, specFileName} = input
    const [coveragePath, actualFileName, extension] = getCoveragePath(
      dir,
      fileName
    )

    try {
      const testPattern = `-- ${specFileName}`

      const command = [
        JEST_BIN,
        '-u',
        '--detectOpenHandles',
        '--maxWorkers=1',
        '--env=node',
        '--collectCoverage=true',
        coveragePath,
        testPattern,
      ].join(' ')
      this.jestChild = execa.command(command, {cwd: dir})
      log('sepx')
      log('Jest start', 'info')
      const result = await this.jestChild
      log('Jest end', 'info')
      this.jestChild = undefined

      return [false, result, actualFileName, extension]
    } catch (e: any) {
      if (e.isCanceled) return [true]
      if (!e.stdout) return [true]
      if (!e.stderr) return [true]

      return [
        false,
        {
          stdout: e.stdout,
          stderr: e.stderr,
        },
        actualFileName,
        extension,
      ]
    }
  }

  parseCoverage(parseCoverageInput: ParseCoverage) {
    const {
      execResult,
      actualFileName,
      fileName,
      extension,
      hasError,
    } = parseCoverageInput
    if (hasError) return {}
    
    const input = cleanAngularLog(execResult)
    const pass = input.stderr.includes('PASS')
    const jestOutputLines = input.stdout.split('\n')

    let foundCoverage = false
    const [lineWithCoverage] = jestOutputLines.filter(line => {
      if (line.includes('% Stmts')) foundCoverage = true

      return foundCoverage && line.includes(`${actualFileName}${extension}`)
    })

    if (lineWithCoverage === undefined) {
      return {
        pass,
        message: SUCCESS_ICON,
      }
    }

    const [
      ,
      statements,
      branch,
      func,
      lines,
      uncovered,
    ] = lineWithCoverage.split('|').map(extractNumber)

    const message = this.getCoverageDiff(
      [statements, branch, func, lines],
      fileName
    )

    return {
      pass,
      message: message === undefined ? 'MESSAGE === undefined' : message,
      uncovered,
    }
  }

  getCoverageDiff(inputs: any[], filePath: string) {
    const fileKey = createFileKey(filePath)
    const firstTime = this.coverageHolder[fileKey] === undefined
    const [statements, branch, func, lines] = inputs.map(toNumber)

    const hash = {
      branch,
      func,
      lines,
      statements,
    }

    const coverageInfo = glue(`
    statements ${STATEMENTS}:
    ${statements} % ${SEPARATOR}
    branches ${BRANCHES}:
    ${branch} % ${SEPARATOR}
    functions ${FUNCTIONS}:
    ${func} % ${SEPARATOR}
    lines ${LINES}:
    ${lines} %
  `)

    if (firstTime) {
      this.coverageHolder[fileKey] = hash

      return coverageInfo
    }

    const statementsDiff = parse(
      statements - this.coverageHolder[fileKey].statements
    )
    const branchDiff = parse(branch - this.coverageHolder[fileKey].branch)
    const funcDiff = parse(func - this.coverageHolder[fileKey].func)
    const linesDiff = parse(lines - this.coverageHolder[fileKey].lines)

    this.coverageHolder[fileKey] = hash

    const message = glue(`
      ${statementsDiff === 0 ? '' : `✍:${maybeWarn(statementsDiff)}`}
      ${branchDiff === 0 ? '' : `🎋:${maybeWarn(branchDiff)}`}
      ${funcDiff === 0 ? '' : `☈:${maybeWarn(funcDiff)}`}
      ${linesDiff === 0 ? '' : `📜:${maybeWarn(linesDiff)}`}
    `)

    const staleCoverageInfo = glue(`
    NO CHANGE ${STALE_SEPARATOR}
    ${STATEMENTS}:
    ${statements} % ${STALE_SEPARATOR}
    ${BRANCHES}:
    ${branch} % ${STALE_SEPARATOR}
    ${FUNCTIONS}:
    ${func} % ${STALE_SEPARATOR}
    ${LINES}:
    ${lines} %
  `)

    return message.trim() === '' ? staleCoverageInfo : `change: ${message}`
  }

  onCancelMessage() {
    if (!this.jestChild) return log('', 'error')
    if (!this.jestChild.cancel) return

    this.jestChild.cancel()
    this.jestChild = undefined
  }

  async onSocketData(messageFromVSCode: any) {
    let parsedMessage: false | Message

    try {
      parsedMessage = JSON.parse(messageFromVSCode.toString())
    } catch (_) {
      parsedMessage = false
    }

    if (parsedMessage === false) {
      return onWrongIncomingMessage(messageFromVSCode.toString())
    }

    debugLog(parsedMessage, 'onSocketData')

    if (parsedMessage.requestCancelation) {
      return this.onCancelMessage()
    }
    if(parsedMessage.fileName.endsWith('.py')){
      await this.onPytestMessage(parsedMessage)
      return
    }
    await this.onJestMessage(parsedMessage)
  }

  start() {
    log(this.initialized ? 'Already initialized' : 'start niketa TDD tool', 'box')
    this.server = createServer(socket => {
      this.initialized = true

      socket.on('data', data => this.onSocketData(data.toString()))

      socket.on('error', err => {
        this.initialized = false
        console.log(err, 'socket.error.niketa.client')
        this.server.close(() => {
          console.log('niketa TDD closed')
          delay(2000).then(() => {
            console.log('niketa TDD will restart')
            this.start()
          })
        })
      })

      this.emit = message => {
        socket.write(JSON.stringify(message))
        socket.pipe(socket)
      }
    })

    this.server.listen(this.port, '127.0.0.1')
  }
}
