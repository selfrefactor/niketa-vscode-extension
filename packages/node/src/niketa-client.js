import execa from 'execa'
import { log } from 'helpers-fn'
import { createServer } from 'net'
import {
  filter,
  glue,
  pass,
  remove,
  tryCatch,
} from 'rambdax'

import { createFileKey } from './utils/create-file-key'
import { isLintOnlyMode } from './utils/is-lint-only-mode'
import { getCoveragePath } from './utils/get-coverage-path'
import { cleanJestOutput } from './utils/clean-jest-output.js'
import { extractConsoleLogs } from './utils/extract-console.logs'
import { getSpecFile } from './utils/get-spec-file.js'

const JEST_BIN = './node_modules/jest/bin/jest.js'
export const ERROR_ICON = '‚ùå'
export const SUCCESS_ICON = 'üê¨'

function cleanAngularLog(x){
  return {
    ...x,
    stderr : remove(/ts-jest\[.+/, x.stderr),
  }
}

function toNumber(x){
  return x === undefined || Number.isNaN(Number(x)) ? 0 : Number(x)
}

function parse(x){
  const result = Math.round(x * 100) / 100

  return parseFloat(`${ result }`)
}

const maybeWarn = x => x < 0 ? `‚ùó${ x }` : x

function extractNumber(text){
  const justText = text.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g,
    '')

  return Number(justText.trim())
}

const defaultEmit = x => console.log(x, 'emit not yet initialized')

const messageSchema = {
  disableLint    : Boolean,
  withLockedFile : Boolean,
  fileName       : String,
  hasWallaby     : Boolean,
}

function isMessageCorrect(message){
  const isCorrect = pass(message)(messageSchema)
  if (!isCorrect){
    log('isMessageCorrect', 'error')

    return false
  }

  return true
}

function getUncoveredMessage(message){
  if (typeof message !== 'string' || !message){
    return
  }

  const uncovered = remove('...', message)

  return message.includes('...') ? `‚õ±${ uncovered }` : `‚òî${ uncovered }`
}

export class NiketaClient{
  constructor(port, emit){
    this.port = port
    this.serverInit = false
    this.coverageHolder = {}
    this.emit = emit === undefined ? defaultEmit : emit
    this.lintFileHolder = undefined
    this.fileHolder = undefined
    this.specFileHolder = undefined
  }

  async onJestMessage(message){
    const { disableLint, fileName, hasWallaby, dir } = message

    if (!isMessageCorrect(message)) return
    if (isLintOnlyMode(fileName)) return this.onLintOnlyMode(fileName)

    const maybeSpecFile = getSpecFile(fileName)
    const { canContinue } = this.markFileForLint({
      maybeSpecFile,
      disableLint,
      hasWallaby,
      fileName,
    })

    if (!canContinue) return
    if (this.stillWaitingForSpec(fileName, dir)) return

    const [
      failure,
      execResult,
      actualFileName,
      extension,
    ] = await this.execJest({
      dir,
      fileName     : this.fileHolder,
      specFileName : this.specFileHolder,
    })

    // if (failure) return 
    process.stderr.write('\nüê¨\n' + execResult.stderr + '\n\n')
    process.stderr.write('\nüê¨\n' + execResult.stdout + '\n\n')

    this.sendToVSCode({
      execResult,
      actualFileName,
      fileName : this.fileHolder,
      extension,
    })

    return true
  }

  sendToVSCode({ execResult, actualFileName, fileName, extension }){
    const hasError =
      execResult.stderr.startsWith('FAIL') ||
      execResult.stderr.includes('ERROR:')
    if (hasError){
      return this.emit({
        firstBarMessage : ERROR_ICON,
        hasDecorations  : false,
      })
    }

    const { pass, message, uncovered } = this.parseCoverage({
      execResult,
      actualFileName,
      fileName,
      extension,
    })
    const { newDecorations, hasDecorations } = this.getNewDecorations({
      execResult,
      fileName,
    })
    const firstBarMessage = pass ? message : ERROR_ICON
    const secondBarMessage = getUncoveredMessage(uncovered)

    this.emit({
      firstBarMessage,
      secondBarMessage,
      hasDecorations,
      newDecorations,
    })
  }

  getNewDecorations({ execResult, fileName }){
    const input = cleanJestOutput(execResult.stdout)
    const [ consoleLogs ] = input.split('----------------------|')
    const newDecorationsData = extractConsoleLogs(consoleLogs)

    if (Object.keys(newDecorationsData).length === 0){
      return { hasDecorations : false }
    }

    const newDecorations = this.evaluateDecorations({
      newDecorationsData,
      fileName,
    })

    return {
      hasDecorations : true,
      newDecorations,
    }
  }

  evaluateDecorations({ newDecorationsData, fileName }){
    const unreliableLogData = []
    const reliableLogData = {}

    const triggerFileHasDecoration = filter((logData, prop) => {
      const okLogData = fileName.endsWith(prop)

      logData.map(({ line, decoration }) => {
        unreliableLogData.push(decoration)
        if (okLogData){
          reliableLogData[ line ] = decoration
        }
      })

      return okLogData
    })(newDecorationsData)

    const correct = Object.keys(triggerFileHasDecoration).length === 1
    const logData = correct ? reliableLogData : unreliableLogData

    return {
      correct,
      logData,
    }
  }

  async execJest({ fileName, dir, specFileName }){
    const [ coveragePath, actualFileName, extension ] = getCoveragePath(dir,
      fileName)

    try {
      const testPattern = `-- ${ specFileName }`

      const command = [
        JEST_BIN,
        '-u',
        '--maxWorkers=1',
        '--env=node',
        '--collectCoverage=true',
        coveragePath,
        testPattern,
      ].join(' ')
      this.jestChild = execa.command(command, { cwd : dir })
      const result = await this.jestChild
      this.jestChild = undefined

      return [ false, result, actualFileName, extension ]
    } catch (e){
      if(e.isCanceled) return [false]
      if(!e.stdout) return [false] 
      if(!e.stderr) return [false]
       
      return [ true, {stdout: e.stdout, stderr: e.stderr},  actualFileName, extension ]
    }
  }

  parseCoverage({ execResult, actualFileName, fileName, extension }){
    const input = cleanAngularLog(execResult)
    const pass = input.stderr.includes('PASS')
    const jestOutputLines = input.stdout.split('\n')

    const [ line ] = jestOutputLines.filter(x =>
      x.includes(`${ actualFileName }${ extension }`))
    
    if (line === undefined){
      return {
        pass,
        message : SUCCESS_ICON,
      }
    }

    const [ , statements, branch, func, lines, uncovered ] = line
      .split('|')
      .map(extractNumber)

    const message = this.getCoverageDiff([ statements, branch, func, lines ], fileName)

    return {
      pass,
      message : message === undefined ? 'MESSAGE === undefined' : message,
      uncovered,
    }
  }

  getCoverageDiff(inputs, filePath){
    const fileKey = createFileKey(filePath)
    const firstTime = this.coverageHolder[ fileKey ] === undefined
    const [ statements, branch, func, lines ] = inputs.map(toNumber)

    const hash = {
      branch,
      func,
      lines,
      statements,
    }

    if (firstTime){
      this.coverageHolder[ fileKey ] = hash

      return glue(`
        üê∞
        st:
        ${ statements }
        br:
        ${ branch }
        fn:
        ${ func }
        lns:
        ${ lines }
      `)
    }

    const statementsDiff = parse(statements - this.coverageHolder[ fileKey ].statements)
    const branchDiff = parse(branch - this.coverageHolder[ fileKey ].branch)
    const funcDiff = parse(func - this.coverageHolder[ fileKey ].func)
    const linesDiff = parse(lines - this.coverageHolder[ fileKey ].lines)

    this.coverageHolder[ fileKey ] = hash

    const message = glue(`
      ${ statementsDiff === 0 ? '' : `‚úç:${ maybeWarn(statementsDiff) }` }
      ${ branchDiff === 0 ? '' : `üéã:${ maybeWarn(branchDiff) }` }
      ${ funcDiff === 0 ? '' : `‚òà:${ maybeWarn(funcDiff) }` }
      ${ linesDiff === 0 ? '' : `üìú:${ maybeWarn(linesDiff) }` }
    `)

    return message.trim() === '' ? '‚õπ' : message
  }

  logError(e, label){
    console.log({
      e,
      label,
    })
  }

  stillWaitingForSpec(fileName, dir){
    const stillWating = !(this.fileHolder && this.specFileHolder)
    const specBelongs = this.fileHolder.startsWith(dir)
    if (stillWating){
      // This happens only until the script receives a correct filepath
      this.debugLog('no specfile', fileName)

      return true
    }
    if (!specBelongs){
      // when we have filepath from previous project but not in the current

      this.debugLog(dir, 'still waiting for testable file in this project')

      return true
    }

    return false
  }

  onLintOnlyMode(fileName){
    console.log('onLintOnlyMode', fileName)
  }

  markFileForLint({ disableLint, fileName, hasWallaby, maybeSpecFile }){
    if (disableLint) return { canContinue : true }

    const allowLint =
      fileName !== this.lintFileHolder && this.lintFileHolder !== undefined

    if (allowLint){
      log(`LINT ${ this.lintFileHolder }`, 'box')
      // whenFileLoseFocus(lintFileHolder, disableLint)
      this.lintFileHolder = fileName
    } else {
      log(`SKIP_LINT ${
        this.lintFileHolder ? this.lintFileHolder : 'initial state'
      }`,
      'box')
    }

    if (hasWallaby){
      this.lintFileHolder = fileName

      this.debugLog(fileName, 'saved for lint later')

      return { canContinue : false }
    }

    if (maybeSpecFile){
      this.fileHolder = fileName
      this.lintFileHolder = fileName
      this.specFileHolder = maybeSpecFile
      this.debugLog(fileName, 'saved for lint later')

      return { canContinue : true }
    }
    this.debugLog(fileName, 'saved for lint later even without spec')

    // Even if the file has no corresponding spec file
    // we keep it for further linting
    this.lintFileHolder = fileName

    return { canContinue : false }
  }

  debugLog(data, label){}

  onCancelMessage({ fileName }){
    console.log('in cancel message', fileName)
    if (!this.jestChild) return
    if (!this.jestChild.cancel) return

    this.jestChild.cancel()
    this.jestChild = undefined
  }

  async onSocketData(messageFromVSCode){
    const parsedMessage = tryCatch(() => JSON.parse(messageFromVSCode.toString()),
      false)()
    if (parsedMessage === false){
      return this.onWrongIncomingMessage(messageFromVSCode.toString())
    }
    if (parsedMessage.requestCancelation){
      return this.onCancelMessage(parsedMessage)
    }
    const result = await this.onJestMessage(parsedMessage)

    return result
  }

  start(){
    if (this.serverInit) return
    this.server = createServer(socket => {
      log('Server created', 'info')
      socket.on('data', data => this.onSocketData(data.toString()))

      this.emit = message => {
        socket.write(JSON.stringify(message))
        socket.pipe(socket)
      }
      this.serverInit = true
    })

    log(`Listen at ${ this.port } for vscode`, 'back')
    this.server.listen(this.port, '127.0.0.1')
  }

  onWrongIncomingMessage(messageFromVSCode){
    return log('Error while parsing messageFromVSCode', 'error')
  }
}

/*
    async onMessageTest(message){
    console.log({ message })
    await delay(4000)
    const testUnreliableData = {
      correct : false,
      logData : [ 'foo', 'foo1', 'foo2', 'foo3' ],
    }

    const newDecorations = {
      correct : true,
      logData : {
        1 : 'foo',
        5 : 'foo1',
        6 : 'foo1',
        8 : 'foo2',
        9 : 'foo3',
      },
    }
    console.log('sending message')
    this.emit({
      newDecorations : testUnreliableData,
      firstStatusBar : 'Keep it up',
    })
  }

  const SHORT_SEPARATOR = repeat('üçÑ', 2).join``
const SEPARATOR = repeat('üç∫', 20).join``

export function debugLog(toLog, label = 'debug log'){
  if (!getter('DEBUG_LOG')) return

  console.log(label, SHORT_SEPARATOR)
  console.log(SEPARATOR)
  console.log(toLog)
  console.log(SEPARATOR)
}

*/
