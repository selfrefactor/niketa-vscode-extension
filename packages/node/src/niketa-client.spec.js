import {resolve} from 'path'
import { NiketaClient } from './niketa-client.js'
jest.setTimeout(10000)

const throwingSyncTest = resolve(__dirname, '../test-data/failed-code/sync.js')
const throwingAsyncTest = resolve(__dirname, '../test-data/failed-code/async.js')

const FAILED_EXPECTATIONS = {
  specHasLogsFile: resolve(__dirname, '../test-data/failed-expectation/spec-has-logs.js'),
  specHasLogsSpec: resolve(__dirname, '../test-data/failed-expectation/spec-has-logs.spec.js'),
  fileHasLogsFile: resolve(__dirname, '../test-data/failed-expectation/file-has-logs.js'),
  fileHasLogsSpec: resolve(__dirname, '../test-data/failed-expectation/file-has-logs.spec.js'),
  bothHaveLogsFile: resolve(__dirname, '../test-data/failed-expectation/both-have-logs.js'),
  bothHaveLogsSpec: resolve(__dirname, '../test-data/failed-expectation/both-have-logs.spec.js')
}

const SUCCESS = {
  specHasLogsFile: resolve(__dirname, '../test-data/success/spec-has-logs.js'),
  specHasLogsSpec: resolve(__dirname, '../test-data/success/spec-has-logs.spec.js'),
  fileHasLogsFile: resolve(__dirname, '../test-data/success/file-has-logs.js'),
  fileHasLogsSpec: resolve(__dirname, '../test-data/success/file-has-logs.spec.js'),
  bothHaveLogsFile: resolve(__dirname, '../test-data/success/both-have-logs.js'),
  bothHaveLogsSpec: resolve(__dirname, '../test-data/success/both-have-logs.spec.js')
}

const testDir = '/home/s/repos/niketa/packages/node'

let niketaClient
const emit = jest.fn()

beforeEach(() => {
  niketaClient = new NiketaClient(3939, emit)
})
afterEach(() => {
  niketaClient = undefined
  emit.mockClear()
})

function generateMessage(input){
  return JSON.stringify({
    disableLint    : false,
    hasWallaby     : false,
    dir            : testDir,
    withLockedFile : false,
    ...input
  })
}


Object.keys(SUCCESS).forEach(testKey => {
    test(`success - ${testKey}`, async () => {
        await niketaClient.onSocketData(generateMessage({
          fileName : SUCCESS[testKey]
        }))
        expect(emit.mock.calls[ 0 ]).toMatchSnapshot()
    })
  }
)

Object.keys(FAILED_EXPECTATIONS).forEach(testKey => {
    test(`failed expectation - ${testKey}`, async () => {
        await niketaClient.onSocketData(generateMessage({
          fileName : FAILED_EXPECTATIONS[testKey]
        }))
        expect(emit.mock.calls[ 0 ]).toMatchSnapshot()
    })
  }
)

test('sync code that throws', async () => {
    await niketaClient.onSocketData(generateMessage({
      fileName : throwingSyncTest
    }))
    expect(emit.mock.calls[ 0 ]).toMatchSnapshot()
})

test('async code that throws', async () => {
    await niketaClient.onSocketData(generateMessage({
      fileName : throwingAsyncTest
    }))
    expect(emit.mock.calls[ 0 ]).toMatchSnapshot()
})

