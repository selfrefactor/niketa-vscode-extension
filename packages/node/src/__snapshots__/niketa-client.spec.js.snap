// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async code that throws 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "4": "1",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - bothHaveLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "2": "2",
        "3": "number",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - bothHaveLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "7": "{ a: 1, b: 2 }",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - fileHasLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "2": "9",
        "3": "number",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - fileHasLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": false,
      "logData": Array [
        "9",
        "number",
      ],
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - specHasLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": false,
      "logData": Array [
        "[ 1, 2, 3 ] { a: 1, b: 2 }",
        "1",
      ],
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`failed expectation - specHasLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "4": "[ 1, 2, 3 ] { a: 1, b: 2 }",
        "5": "1",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - bothHaveLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 0 br: 0 fn: 0 lns: 0",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "2": "2",
        "3": "number",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - bothHaveLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 0 br: 0 fn: 0 lns: 0",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "7": "{ a: 1, b: 2 }",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - fileHasLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 0 br: 0 fn: 0 lns: 0",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "2": "9",
        "3": "number",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - fileHasLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 0 br: 0 fn: 0 lns: 0",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": false,
      "logData": Array [
        "9",
        "number",
      ],
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - specHasLogsFile 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 100 br: 50 fn: 100 lns: 100",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": false,
      "logData": Array [
        "[ 1, 2, 3 ] { a: 1, b: 2 }",
        "1",
      ],
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`success - specHasLogsSpec 1`] = `
Array [
  Object {
    "firstBarMessage": "üê∞ st: 100 br: 50 fn: 100 lns: 100",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "4": "[ 1, 2, 3 ] { a: 1, b: 2 }",
        "5": "1",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;

exports[`sync code that throws 1`] = `
Array [
  Object {
    "firstBarMessage": "‚ùå",
    "hasDecorations": true,
    "newDecorations": Object {
      "correct": true,
      "logData": Object {
        "2": "1",
      },
    },
    "secondBarMessage": undefined,
  },
]
`;
